### PR TITLE
Slightly simpler / faster variant

### PR DESCRIPTION
**Pull request by WebReflection**
As discussed in https://github.com/felixfbecker/node-sql-template-strings/issues/102

### CODE REVIEW

File: test/unit.js [hunk 1]
```diff
@@ -70,7 +70,6 @@ describe('SQL', () => {
       assert.equal(statement.text, 'SELECT * FROM table WHERE column = $1 ORDER BY other_column')
       assert.strictEqual(statement.query, 'SELECT * FROM table WHERE column = $1 ORDER BY other_column')
       assert.strictEqual(statement.values, undefined)
-      assert.strictEqual('values' in statement, false)
```
Comment (felixfbecker) [inline test/unit.js]: Why did you remove this assertion? I vaguely remember that this was added intentionally because one of the DB drivers used a `for in` loop to iterate over options. So it is critical that the property is enumerable. Not passing this assertion would mean a regression and break compatibility. This is why there is an `Object.defineProperty()` call in the version on master.

File: index.js [hunk 2]
```diff
@@ -1,89 +1,91 @@
...
+      return this
+    },
+
+    // Properties
+
+    /** Default bound value as false */
+    bound: false,
+
+    /** Default name as empty string */
+    name: '',
```
Comment (felixfbecker) [inline index.js]: Shouldn't a statement be not named at all by default (`undefined`)?

File: index.js [hunk 3]
```diff
@@ -1,89 +1,91 @@
...
+
+    /** Initial strings */
+    strings,
```
Comment (felixfbecker) [inline index.js]: These used to be private, is there a reason to expose them?

File: test/unit.js (hunk 1 above)
Comment (WebReflection) [reply test/unit.js]: like I've said, adding and removing properties to an object makes the object slow.
If you understand the changes I've made, you'll realize properties are getters and returned accordingly to the `bound` state.
If this breaks anywhere, it should be stated where, which version, and why.

I can move these two outside as non enumerable, if that solves the `for in` issue, but that won't solve `in` ... which is a different story.

So, in few words, are you asking for non enumerables, or for runtime deleted properties that makes this whole optimization useless?

File: index.js (hunk 2 above)
Comment (WebReflection) [reply index.js]: same as https://github.com/felixfbecker/node-sql-template-strings/pull/103#discussion_r260771148

this whole refactoring is to have specialized, optimized, V8 internal class with an object that never deletes, or change, its fields.

File: index.js (hunk 3 above)
Comment (WebReflection) [reply index.js]: strings property has never been private in your code, you access it on `append` already, and you set it as configurable, enumerable, and writable, within the `constructor`, same as here.

Comment (felixfbecker) [reply index.js]: I consider a proeprty "private" when it is not part of the public (documented) API. `append()` is part of the class so it can use private properties. Check the documentation which mentions it as private: http://node-sql-template-strings.surge.sh/classes/sqlstatement.html#strings
Before in a class we needed it as a property, but now `strings` is in scope, so this wouldn't be needed anymore.

File: index.js (hunk 2 above)
Comment (felixfbecker) [reply index.js]: I am just afraid of breaking a DB driver that may check for a name like this:

```js
if (typeof options.name === 'string')
if (options.name !== undefined)
```

Comment (WebReflection) [reply index.js]: that driver will fail when the object changes at runtime though, right? 'cause if it has a different usage for when the name is there or not, mutating the object won't solve the root cause of possible issues.

File: test/unit.js (hunk 1 above)
Comment (felixfbecker) [reply test/unit.js]: I understand the changes you made. The problem is the integration tests cannot detect whether a query was actually executed as a prepared statement or not, so you'll have to verify that manually.
It seems like the only constraint though is that the property is enumerable: https://sourcegraph.com/github.com/mysqljs/mysql@2a9239871fb24515485476e7c4d1deb941f525e9/-/blob/lib/Connection.js#L39-41
I couldn't find another gotcha from reading the DB driver code. Still, it should be tested.

File: index.js (hunk 3 above)
Comment (WebReflection) [reply index.js]: it's still needed via `append`, so either I use a `WeakMap` to have real privates, or it stays documented as private as it was before.

This literally doesn't change anything to the object, its property descriptor, or the accessibility of `strings`. I could state private as comment, but is exactly as "_private_" as before.

File: index.js (hunk 2 above)
Comment (WebReflection) [reply index.js]: also, as stated in here https://github.com/felixfbecker/node-sql-template-strings/blob/master/index.d.ts#L24-L27 the `name` is a string by specs/API/definition

File: index.js (hunk 3 above)
Comment (WebReflection) [reply index.js]: As extra note: accessing `strings` is necessary for backward compatibility, so there's literally no point, since it's already documented as private, to not carry along the property s it was done before.

**Comment by WebReflection** [comment]
To clarify: this is all; about **not** having object fields/properties **ever** added or removed at runtime.

If you don't want that, we can close this PR since there's nothing else to do here 👋

File: test/unit.js (hunk 1 above)
Comment (WebReflection) [reply test/unit.js]: That deriver overwrites `.values` regardless of the `for/in` https://sourcegraph.com/github.com/mysqljs/mysql@2a9239871fb24515485476e7c4d1deb941f525e9/-/blob/lib/Connection.js#L46 and in your current code, you set the `values` as writable, enumerable, and configurable, in the constructor.

So, basically, nothing changes at all for that driver: it was setting `options.values = sql.values` regardless even before these changes.

Comment (WebReflection) [reply test/unit.js]: TL;DR if mysql/I driver never uses `useBind(true)` this change won't have any undesired side-effect. If it does, I don't know how that driver could work.

If I make both `values` and `bind` non enumerable, the `options` will never have values in case the passed one is a callback, and I don't know how that would behave.

I understand your concerns though, but I'd like to see a practical failure with current object.

File: index.js (hunk 2 above)
Comment (felixfbecker) [reply index.js]: No, cause you're only changing the object before you pass it to the driver.
The `name` TS declaration was written before TS had support for optional class properties, it was `undefined` at runtime if not set I think.

File: index.js (hunk 3 above)
Comment (felixfbecker) [reply index.js]: Why would WeakMaps be needed? Isn't `strings` in scope for each new instance?

File: test/unit.js (hunk 1 above)
Comment (felixfbecker) [reply test/unit.js]: Did you test it at runtime (with the other drivers too)?

File: index.js (hunk 2 above)
Comment (WebReflection) [reply index.js]: this is similar issue we have with other fields ... if you have `name in`

File: index.js (hunk 3 above)
Comment (WebReflection) [reply index.js]: `append(object)` ... how do you get `object.strings` there ?

File: test/unit.js (hunk 1 above)
Comment (WebReflection) [reply test/unit.js]: nope, but I guess that'd be an improvement for this library in general ... I mean, having tests to validate noting breaks, with all supported drivers.

Comment (WebReflection) [reply test/unit.js]: Just to clarify, I wasn't planing to push those tests so, I wonder if we're moving forward with this PR ('cause I have more coming to improve `append` etc)

**Comment by WebReflection** [comment]
OK, I'll close this since it's not moving and it won't bring much improvements on the real-world.

**Comment by felixfbecker** [comment]
It's been 2 days. I just didn't have time to review again.
